compiler/sem/passes.nim

```diff
@@ -84,13 +85,26 @@ proc closePasses(graph: ModuleGraph; a: var TPassContextArray) =
       m = graph.passes[i].close(graph, a[i], m)
     a[i] = nil                # free the memory here
 
-proc processTopLevelStmt(graph: ModuleGraph, n: PNode, a: var TPassContextArray): bool =
-  # this implements the code transformation pipeline
-  var m = n
-  for i in 0..<graph.passes.len:
+proc processTopLevelStmt(
+    graph: ModuleGraph,
+    toProcess: PNode,
+    a: var TPassContextArray): bool =
+  ## Main processing pipeline entry point - accepts a toplevel node,
+  ## collection of pass contexts and main module graph which defines passes
+  ## to use.
+  # First step works with the base tree as entry point
+  var processingResult = toProcess
+  for i in 0 ..< graph.passes.len:
     if not isNil(graph.passes[i].process):
-      m = graph.passes[i].process(a[i], m)
-      if isNil(m): return false
+      # Iterate over all processing passes, re-assigning the evaluation
+      # results each time.
+      processingResult = graph.passes[i].process(a[i], processingResult)
+      if isNil(processingResult):
+        # QUESTION this is an error/fatal diagnostic or some parts of the
```

Comment:
I think sem and backend can both end up doing this especially through legacy code paths.